jobrunner/create_or_update_jobs.py

```diff
@@ -0,0 +1,247 @@
+"""
+This module provides a single public entry point `create_or_update_jobs`.
+
+It handles all logic connected with creating or updating Jobs in response to
+JobRequests. This includes fetching the code with git, validating the project
+and doing the necessary dependency resolution.
+"""
+from pathlib import Path
+import re
+import time
+
+from . import config
+from .database import transaction, insert, exists_where, find_where
+from .git import read_file_from_repo, get_sha_from_remote_ref, GitError
+from .project import (
+    parse_and_validate_project_file,
+    get_action_specification,
+    get_all_actions,
+    ProjectValidationError,
+)
+from .models import Job, SavedJobRequest, State
+from .manage_jobs import action_has_successful_outputs
+
+
+class JobRequestError(Exception):
+    pass
+
+
+class NothingToDoError(JobRequestError):
+    pass
+
+
+def create_or_update_jobs(job_request):
+    """
+    Create or update Jobs in response to a JobRequest
+
+    Note that where there is an error with the JobRequest it will create a
+    single, failed job with the error details rather than raising an exception.
+    This allows the error to be synced back to the job-server where it can be
+    displayed to the user.
+    """
+    if not related_jobs_exist(job_request):
+        try:
+            create_jobs(job_request)
+        except (GitError, ProjectValidationError, JobRequestError) as e:
+            create_failed_job(job_request, e)
+    else:
+        # TODO: think about what sort of updates we want to support
+        # I think these are probably limited to:
+        #  * cancel any pending jobs
+        #  * cancel any pending jobs and kill any running ones
+        #  * update the target commit SHA for any pending jobs (although cancel and
+        #    resubmit would also work for this and would probably be simpler)
+        #
+        # This could be implemented by adding a boolean `cancel` column to the
+        # job table. The run loop would check the value of this each time it
+        # checks the state of the job and if it's set it would either call
+        # `docker kill` on it (if it's running) or move it immediately to the
+        # FAILED state (if it's still pending).
+        pass
+
+
+def related_jobs_exist(job_request):
+    return exists_where(Job, job_request_id=job_request.id)
+
+
+def create_jobs(job_request):
+    validate_job_request(job_request)
+    # In future I expect the job-server to only ever supply commits and so this
+    # branch resolution will be redundant
+    if not job_request.commit:
+        job_request.commit = get_sha_from_remote_ref(
+            job_request.repo_url, job_request.branch
+        )
```

Comment:
This should be redundant now that job-server provides a commit with each JobRequest??